# Github

## Import datasets and github
* Clone chestxray dataset from the github link https://github.com/ieee8023/covid-chestxray-dataset.git

* RSNA Pneumonia Detection Challenge dataset
https://www.kaggle.com/c/rsna-pneumonia-detection-challenge and unzip rsna_dataset here

In [1]:
! git clone https://github.com/ieee8023/covid-chestxray-dataset.git

Cloning into 'covid-chestxray-dataset'...
Updating files:  60% (708/1174)
Updating files:  61% (717/1174)
Updating files:  62% (728/1174)
Updating files:  63% (740/1174)
Updating files:  64% (752/1174)
Updating files:  65% (764/1174)
Updating files:  66% (775/1174)
Updating files:  67% (787/1174)
Updating files:  68% (799/1174)
Updating files:  69% (811/1174)
Updating files:  70% (822/1174)
Updating files:  71% (834/1174)
Updating files:  72% (846/1174)
Updating files:  73% (858/1174)
Updating files:  74% (869/1174)
Updating files:  75% (881/1174)
Updating files:  76% (893/1174)
Updating files:  77% (904/1174)
Updating files:  78% (916/1174)
Updating files:  79% (928/1174)
Updating files:  80% (940/1174)
Updating files:  81% (951/1174)
Updating files:  81% (959/1174)
Updating files:  82% (963/1174)
Updating files:  83% (975/1174)
Updating files:  84% (987/1174)
Updating files:  85% (998/1174)
Updating files:  86% (1010/1174)
Updating files:  87% (1022/1174)
Updating files:  88% (1034/1

In [2]:
! git clone https://github.com/agchung/Figure1-COVID-chestxray-dataset.git

Cloning into 'Figure1-COVID-chestxray-dataset'...


In [3]:
! git clone https://github.com/agchung/Actualmed-COVID-chestxray-dataset.git

Cloning into 'Actualmed-COVID-chestxray-dataset'...
Updating files:  23% (57/240)
Updating files:  24% (58/240)
Updating files:  25% (60/240)
Updating files:  26% (63/240)
Updating files:  27% (65/240)
Updating files:  28% (68/240)
Updating files:  29% (70/240)
Updating files:  30% (72/240)
Updating files:  31% (75/240)
Updating files:  32% (77/240)
Updating files:  33% (80/240)
Updating files:  34% (82/240)
Updating files:  35% (84/240)
Updating files:  36% (87/240)
Updating files:  37% (89/240)
Updating files:  38% (92/240)
Updating files:  39% (94/240)
Updating files:  40% (96/240)
Updating files:  40% (98/240)
Updating files:  41% (99/240)
Updating files:  42% (101/240)
Updating files:  43% (104/240)
Updating files:  44% (106/240)
Updating files:  45% (108/240)
Updating files:  46% (111/240)
Updating files:  47% (113/240)
Updating files:  48% (116/240)
Updating files:  49% (118/240)
Updating files:  50% (120/240)
Updating files:  51% (123/240)
Updating files:  52% (125/240)
Updatin

In [4]:
! git clone https://github.com/IliasPap/COVIDNet.git

Cloning into 'COVIDNet'...


In [ ]:
COPY_FILE = True


# Imports

In [1]:
import numpy as np
import pandas as pd
import os
import random 
from shutil import copyfile
import pydicom as dicom
import cv2
import os

# Parameters

In [2]:
# set parameters here
savepath = '../Data'
#train = '../Data/train'
#test = '../Data/test'

seed = 0
np.random.seed(seed) # Reset the seed so all runs are the same.
random.seed(seed)
MAXVAL = 255  # Range [0 255]

## Paths

In [3]:
# path to covid-19 dataset from https://github.com/ieee8023/covid-chestxray-dataset
cohen_imgpath = '../Original Data/covid-chestxray-dataset/images' 
cohen_csvpath = '../Original Data/covid-chestxray-dataset/metadata.csv'

# path to covid-19 dataset from https://github.com/agchung/Figure1-COVID-chestxray-dataset
fig1_imgpath = '../Original Data/Figure1-COVID-chestxray-dataset/images'
fig1_csvpath = '../Original Data/Figure1-COVID-chestxray-dataset/metadata.csv'

# path to covid-19 dataset from https://github.com/agchung/Actualmed-COVID-chestxray-dataset
actmed_imgpath = '../Original Data/Actualmed-COVID-chestxray-dataset/images'
actmed_csvpath = '../Original Data/Actualmed-COVID-chestxray-dataset/metadata.csv'

# path to viral pneumonia dataset from https://www.kaggle.com/tawsifurrahman/covid19-radiography-database
sirm_imgpath = '../Original Data/COVID-19_Radiography_Dataset/Viral Pneumonia'
sirm_csvpath = '../Original Data/COVID-19_Radiography_Dataset/Viral Pneumonia.metadata.xlsx'

# path to normal dataset from https://www.kaggle.com/tawsifurrahman/covid19-radiography-database
sirm_imgpath_norm = '../Original Data/COVID-19_Radiography_Dataset/Normal'
sirm_csvpath_norm = '../Original Data/COVID-19_Radiography_Dataset/Normal.metadata.xlsx'

In [4]:
# path to https://www.kaggle.com/c/rsna-pneumonia-detection-challenge
rsna_datapath = '../Original Data/rsna-pneumonia-detection-challenge'
# get all the normal from here
rsna_csvname = 'stage_2_detailed_class_info.csv' 
# get all the 1sfrom here since 1 indicate pneumonia
# found that images that aren't pneunmonia and also not normal are classified as 0s
rsna_csvname2 = 'stage_2_train_labels.csv' 
rsna_imgpath = 'stage_2_train_images'

In [5]:
# parameters for COVIDx dataset
train = []
test = []
test_count = {'normal': 0, 'pneumonia': 0}
train_count = {'normal': 0, 'pneumonia': 0}

mapping = dict()
#mapping['COVID-19'] = 'COVID-19'
mapping['NORMAL'] = 'normal'
mapping['SARS'] = 'pneumonia'
mapping['MERS'] = 'pneumonia'
mapping['Streptococcus'] = 'pneumonia'
mapping['Klebsiella'] = 'pneumonia'
mapping['Chlamydophila'] = 'pneumonia'
mapping['Legionella'] = 'pneumonia'
mapping['Viral Pneumonia'] = 'pneumonia'
mapping['Lung Opacity'] = 'pneumonia'
mapping['Pneumonia'] = 'pneumonia'
mapping['1'] = 'pneumonia'
mapping['0'] = 'normal'
mapping['Normal'] = 'normal'

In [6]:
# train/test split
split = 0.1

# to avoid duplicates
patient_imgpath = {}

In [7]:
cohen_csv = pd.read_csv(cohen_csvpath, nrows=None)
#idx_pa = csv["view"] == "PA"  # Keep only the PA view
views = ["PA", "AP", "AP Supine", "AP semi erect", "AP erect"]
cohen_idx_keep = cohen_csv.view.isin(views)
cohen_csv = cohen_csv[cohen_idx_keep]

fig1_csv = pd.read_csv(fig1_csvpath, encoding='ISO-8859-1', nrows=None)
actmed_csv = pd.read_csv(actmed_csvpath, nrows=None)

sirm_csv = pd.read_excel(sirm_csvpath)

sirm_csv_norm = pd.read_excel(sirm_csvpath_norm)

In [8]:
# get non-COVID19 viral, bacteria, and COVID-19 infections from covid-chestxray-dataset, figure1 and actualmed
# stored as patient id, image filename and label
filename_label = {'normal': [], 'pneumonia': []}
count = {'normal': 0, 'pneumonia': 0}
covid_ds = {'cohen': [], 'fig1': [], 'actmed': [], 'sirm': [], 'sirm_norm': []}

for index, row in cohen_csv.iterrows():
    f = row['finding'].split('/')[0]
    if f in mapping: # 
        count[mapping[f]] += 1
        entry = [str(row['patientid']), row['filename'], mapping[f], 'cohen']
        filename_label[mapping[f]].append(entry)
        if mapping[f] == 'COVID-19':
            covid_ds['cohen'].append(str(row['patientid']))
        
for index, row in fig1_csv.iterrows():
    if not str(row['finding']) == 'nan':
        f = row['finding'].split(',')[0] # take the first finding
        if f in mapping: # 
            count[mapping[f]] += 1
            if os.path.exists(os.path.join(fig1_imgpath, row['patientid'] + '.jpg')):
                entry = [row['patientid'], row['patientid'] + '.jpg', mapping[f], 'fig1']
            elif os.path.exists(os.path.join(fig1_imgpath, row['patientid'] + '.png')):
                entry = [row['patientid'], row['patientid'] + '.png', mapping[f], 'fig1']
            filename_label[mapping[f]].append(entry)
            if mapping[f] == 'COVID-19':
                covid_ds['fig1'].append(row['patientid'])

for index, row in actmed_csv.iterrows():
    if not str(row['finding']) == 'nan':
        f = row['finding'].split(',')[0]
        if f in mapping:
            count[mapping[f]] += 1
            entry = [row['patientid'], row['imagename'], mapping[f], 'actmed']
            filename_label[mapping[f]].append(entry)
            if mapping[f] == 'COVID-19':
                covid_ds['actmed'].append(row['patientid'])
    
sirm = set(sirm_csv['URL'])
cohen = set(cohen_csv['url'])
#discard = ['100', '101', '102', '103', '104', '105',  '110', '111', '112', '113', '122', '123', '124', '125', '126', '217']

for idx, row in sirm_csv.iterrows():
    patientid = row['FILE NAME']
    if row['URL'] not in cohen:
        key = patientid.split('-')[0]
        if key in mapping:
            count[mapping[key]] += 1
            imagename = patientid + '.' + row['FORMAT'].lower()
            if not os.path.exists(os.path.join(sirm_imgpath, imagename)):
                continue
            entry = [patientid, imagename, mapping[key], 'sirm']
            filename_label[mapping[key]].append(entry)
            covid_ds['sirm'].append(patientid)
            
for idx, row in sirm_csv_norm.iterrows():
    patientid = row['FILE NAME']
    if row['URL'] not in cohen:
        key = patientid.split('-')[0]
        if key in mapping:
            count[mapping[key]] += 1
            imagename = patientid + '.' + row['FORMAT'].lower()
            entry = [patientid, imagename, mapping[key], 'sirm_norm']
            filename_label[mapping[key]].append(entry)
            covid_ds['sirm_norm'].append(patientid)

print('Data distribution from covid datasets:')
print(count)

Data distribution from covid datasets:
{'normal': 10192, 'pneumonia': 2015}


Data distribution from covid datasets:
{'normal': 10192, 'pneumonia': 2015, 'COVID-19': 93}

In [11]:
filename_label.keys()

dict_keys(['normal', 'pneumonia'])

In [9]:
a = np.array(filename_label['pneumonia'])


names, counts = np.unique(a[:,0], return_counts = True)

our_test_patients = []
for i in range(len(counts)):
    if counts[i] > 1:
        our_test_patients.append(names[i])
print(our_test_patients)

['10', '101', '112', '113', '114', '115', '116', '117', '13', '132', '139', '142', '143', '144', '15', '150', '160', '163', '164', '167', '168', '169', '17', '171', '173', '178', '179', '183', '185', '187', '19', '190', '191', '192', '196', '197', '198', '2', '205', '206', '219', '221', '222', '225', '226', '228', '230', '235', '243', '244', '245', '246', '249', '250', '252', '253', '254', '263', '264', '272', '277', '278', '279', '282', '283', '284', '288', '289', '290', '291', '293', '294', '299', '3', '300', '303', '305', '31', '310', '313a', '314', '315', '316', '318', '320', '321', '324b', '326a', '331b', '332', '337', '338', '339', '342', '344', '347', '350a', '357', '359', '36', '360b', '362', '364', '365', '368', '369', '37', '379', '380', '382', '383', '396', '399', '4', '403', '404', '410', '412', '413', '416', '420', '423b', '425c', '425d', '426a', '426b', '446', '447', '451', '462', '471', '475', '477', '479', '51', '57', '58', '59', '6', '67', '68', '69', '7', '70', '71', 

In [12]:
# add covid-chestxray-dataset, figure1 and actualmed into COVIDx dataset
# since these datasets don't have test dataset, split into train/test by patientid
# for covid-chestxray-dataset:
# patient 8 is used as non-COVID19 viral test
# patient 31 is used as bacterial test
# patients 19, 20, 36, 42, 86 are used as COVID-19 viral test
# for figure 1:
# patients 24, 25, 27, 29, 30, 32, 33, 36, 37, 38

ds_imgpath = {'cohen': cohen_imgpath, 'fig1': fig1_imgpath, 'actmed': actmed_imgpath, 'sirm': sirm_imgpath,
              'sirm_norm': sirm_imgpath_norm }

for key in filename_label.keys():
    arr = np.array(filename_label[key])
    if arr.size == 0:
        continue
    # split by patients
    # num_diff_patients = len(np.unique(arr[:,0]))
    # num_test = max(1, round(split*num_diff_patients))
    # select num_test number of random patients
    # random.sample(list(arr[:,0]), num_test)
    if key == 'pneumonia':
        test_patients = ['8', '31'] + our_test_patients
    elif key == 'COVID-19':
        
        test_patients = ['19', '20', '36', '42', '86', 
                         '94', '97', '117', '132', 
                         '138', '144', '150', '163', '169', '174', '175', '179', '190', '191'
                         'COVID-00024', 'COVID-00025', 'COVID-00026', 'COVID-00027', 'COVID-00029',
                         'COVID-00030', 'COVID-00032', 'COVID-00033', 'COVID-00035', 'COVID-00036',
                         'COVID-00037', 'COVID-00038',
                         'ANON24', 'ANON45', 'ANON126', 'ANON106', 'ANON67',
                         'ANON153', 'ANON135', 'ANON44', 'ANON29', 'ANON201', 
                         'ANON191', 'ANON234', 'ANON110', 'ANON112', 'ANON73', 
                         'ANON220', 'ANON189', 'ANON30', 'ANON53', 'ANON46',
                         'ANON218', 'ANON240', 'ANON100', 'ANON237', 'ANON158',
                         'ANON174', 'ANON19', 'ANON195',
                         'COVID-19(119)', 'COVID-19(87)', 'COVID-19(70)', 'COVID-19(94)', 
                         'COVID-19(215)', 'COVID-19(77)', 'COVID-19(213)', 'COVID-19(81)', 
                         'COVID-19(216)', 'COVID-19(72)', 'COVID-19(106)', 'COVID-19(131)', 
                         'COVID-19(107)', 'COVID-19(116)', 'COVID-19(95)', 'COVID-19(214)', 
                         'COVID-19(129)']
    else: 
        test_patients = []
    print('Key: ', key)
    print('Test patients: ', test_patients)
    # go through all the patients
    for patient in arr:
        if patient[0] not in patient_imgpath:
            patient_imgpath[patient[0]] = [patient[1]]
        else:
            if patient[1] not in patient_imgpath[patient[0]]:
                patient_imgpath[patient[0]].append(patient[1])
                # print('THERE IS A WAY')
            else:
                continue  # skip since image has already been written
        if patient[0] in test_patients:
            if patient[3] == 'sirm' or patient[3] == 'sirm_norm':
                image = cv2.imread(os.path.join(ds_imgpath[patient[3]], patient[1]))
                gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
                patient[1] = patient[1].replace(' ', '')
                cv2.imwrite(os.path.join(savepath, 'test', key, patient[1]), gray)
            else:
                copyfile(os.path.join(ds_imgpath[patient[3]], patient[1]), os.path.join(savepath, 'test', key, patient[1]))
            test.append(patient)
            test_count[patient[2]] += 1
        else:
            if patient[3] == 'sirm' or patient[3] == 'sirm_norm':
                image = cv2.imread(os.path.join(ds_imgpath[patient[3]], patient[1]))
                gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
                patient[1] = patient[1].replace(' ', '')
                cv2.imwrite(os.path.join(savepath, 'train', key, patient[1]), gray)
            else:
                copyfile(os.path.join(ds_imgpath[patient[3]], patient[1]), os.path.join(savepath, 'train', key, patient[1]))
            train.append(patient)
            train_count[patient[2]] += 1

print('test count: ', test_count)
print('train count: ', train_count)

Key:  normal
Test patients:  []
Key:  pneumonia
Test patients:  ['8', '31', '10', '101', '112', '113', '114', '115', '116', '117', '13', '132', '139', '142', '143', '144', '15', '150', '160', '163', '164', '167', '168', '169', '17', '171', '173', '178', '179', '183', '185', '187', '19', '190', '191', '192', '196', '197', '198', '2', '205', '206', '219', '221', '222', '225', '226', '228', '230', '235', '243', '244', '245', '246', '249', '250', '252', '253', '254', '263', '264', '272', '277', '278', '279', '282', '283', '284', '288', '289', '290', '291', '293', '294', '299', '3', '300', '303', '305', '31', '310', '313a', '314', '315', '316', '318', '320', '321', '324b', '326a', '331b', '332', '337', '338', '339', '342', '344', '347', '350a', '357', '359', '36', '360b', '362', '364', '365', '368', '369', '37', '379', '380', '382', '383', '396', '399', '4', '403', '404', '410', '412', '413', '416', '420', '423b', '425c', '425d', '426a', '426b', '446', '447', '451', '462', '471', '475', '47

Key:  normal
Test patients:  []
Key:  pneumonia
Test patients:  ['8', '31', '10', '101', '112', '113', '114', '115', '116', '117', '13', '132', '139', '142', '143', '144', '15', '150', '160', '163', '164', '167', '168', '169', '17', '171', '173', '178', '179', '183', '185', '187', '19', '190', '191', '192', '196', '197', '198', '2', '205', '206', '219', '221', '222', '225', '226', '228', '230', '235', '243', '244', '245', '246', '249', '250', '252', '253', '254', '263', '264', '272', '277', '278', '279', '282', '283', '284', '288', '289', '290', '291', '293', '294', '299', '3', '300', '303', '305', '31', '310', '313a', '314', '315', '316', '318', '320', '321', '324b', '326a', '331b', '332', '337', '338', '339', '342', '344', '347', '350a', '357', '359', '36', '360b', '362', '364', '365', '368', '369', '37', '379', '380', '382', '383', '396', '399', '4', '403', '404', '410', '412', '413', '416', '420', '423b', '425c', '425d', '426a', '426b', '446', '447', '451', '462', '471', '475', '477', '479', '51', '57', '58', '59', '6', '67', '68', '69', '7', '70', '71', '73', '86', '87', '9', '91', '92', '95', '96', '97']
Key:  COVID-19
Test patients:  ['19', '20', '36', '42', '86', '94', '97', '117', '132', '138', '144', '150', '163', '169', '174', '175', '179', '190', '191COVID-00024', 'COVID-00025', 'COVID-00026', 'COVID-00027', 'COVID-00029', 'COVID-00030', 'COVID-00032', 'COVID-00033', 'COVID-00035', 'COVID-00036', 'COVID-00037', 'COVID-00038', 'ANON24', 'ANON45', 'ANON126', 'ANON106', 'ANON67', 'ANON153', 'ANON135', 'ANON44', 'ANON29', 'ANON201', 'ANON191', 'ANON234', 'ANON110', 'ANON112', 'ANON73', 'ANON220', 'ANON189', 'ANON30', 'ANON53', 'ANON46', 'ANON218', 'ANON240', 'ANON100', 'ANON237', 'ANON158', 'ANON174', 'ANON19', 'ANON195', 'COVID-19(119)', 'COVID-19(87)', 'COVID-19(70)', 'COVID-19(94)', 'COVID-19(215)', 'COVID-19(77)', 'COVID-19(213)', 'COVID-19(81)', 'COVID-19(216)', 'COVID-19(72)', 'COVID-19(106)', 'COVID-19(131)', 'COVID-19(107)', 'COVID-19(116)', 'COVID-19(95)', 'COVID-19(214)', 'COVID-19(129)']
test count:  {'normal': 0, 'pneumonia': 415, 'COVID-19': 44}
train count:  {'normal': 10192, 'pneumonia': 1600, 'COVID-19': 49}

In [13]:
csv_normal = pd.read_csv(os.path.join(rsna_datapath, rsna_csvname), nrows=None)
csv_pneu = pd.read_csv(os.path.join(rsna_datapath, rsna_csvname2), nrows=None)
patients = {'normal': [], 'pneumonia': []}

for index, row in csv_normal.iterrows():
    if row['class'] == 'Normal':
        patients['normal'].append(row['patientId'])

for index, row in csv_pneu.iterrows():
    if int(row['Target']) == 1:
        patients['pneumonia'].append(row['patientId'])

In [13]:
len(csv_pneu)

30227

In [14]:
len(patients['pneumonia'])

9555

In [15]:
len(patients['normal'])

8851

In [16]:
names, counts = np.unique(patients['pneumonia'], return_counts = True)
np.unique(counts)

array([1, 2, 3, 4], dtype=int64)

In [14]:
for key in patients.keys():
    arr = np.array(patients[key])
    if arr.size == 0:
        continue
    # split by patients 
    # num_diff_patients = len(np.unique(arr))
    # num_test = max(1, round(split*num_diff_patients))
    test_patients = np.load('../Original Data/COVIDNet/data/rsna_test_patients_{}.npy'.format(key)) # random.sample(list(arr), num_test), download the .npy files from the repo.
    # np.save('rsna_test_patients_{}.npy'.format(key), np.array(test_patients))
    for patient in arr:
        if patient not in patient_imgpath:
            patient_imgpath[patient] = [patient]
        else:
            continue  # skip since image has already been written
                
        ds = dicom.dcmread(os.path.join(rsna_datapath, rsna_imgpath, patient + '.dcm'))
        pixel_array_numpy = ds.pixel_array
        imgname = patient + '.png'
        if patient in test_patients:
            cv2.imwrite(os.path.join(savepath, 'test', key, imgname), pixel_array_numpy)
            test.append([patient, imgname, key, 'rsna'])
            test_count[key] += 1
        else:
            cv2.imwrite(os.path.join(savepath, 'train', key, imgname), pixel_array_numpy)
            train.append([patient, imgname, key, 'rsna'])
            train_count[key] += 1

print('test count: ', test_count)
print('train count: ', train_count)

test count:  {'normal': 885, 'pneumonia': 1004}
train count:  {'normal': 18158, 'pneumonia': 7023}


In [31]:
test_patients = np.load('../Original Data/COVIDNet/data/rsna_test_patients_{}.npy'.format('normal'))
len(test_patients)

885

In [41]:
test_patients = np.load('../Original Data/COVIDNet/data/rsna_test_patients_{}.npy'.format('pneumonia'))
len(test_patients)

601